### DISCUSSION TITLE
RFC: Use Sharp for image optimization if installed

### DISCUSSION THREAD

**Discussion by atcastle** [vercel/next.js#27073] [RFC]
**Update:** [This has shipped](https://nextjs.org/docs/api-reference/next/image#built-in-loaders)! Run `npm install next@latest sharp` to start using Sharp as the default.

---

### What I'm proposing:

1) Allow the Next.js image optimizer to use Sharp (libvips) for it’s image transformations, if that library is already present on the server (i.e. installed by the user)
2) If Next is started in production mode (next start), uses the built-in image optimizer, and Sharp is not present on the server, log a console warning suggesting that the user install Sharp.

### Why I'm proposing this

[Sharp is a lot faster than Squoosh for this use case](https://github.com/lovell/sharp/issues/2643). This fact has been [noticed by the community](https://github.com/vercel/next.js/issues/23637) since Next.js moved from Sharp to Squoosh, and is an ongoing source of complaints. Some users say this is preventing them from using the image component entirely.

We have no reason to believe that Squoosh will be competitive with Sharp for this use case in the near future. 

The two reasons Next.js switched to Squoosh are both addressed by this solution:
1) For a period, Sharp caused major issues when used on a new M1 Macbook. This issue is now resolved. [Sharp can be used on M1 Macbooks](https://sharp.pixelplumbing.com/install#apple-m1).
2) Sharp slowed down initial Next install times substantially, because it is a large dependency and in some ([but not most](https://sharp.pixelplumbing.com/install#prebuilt-binaries)) cases, libvips and/or sharp must be compiled locally, which adds additional build time. Since many users will not use the built-in image optimizer for production image serving, this is an unacceptable cost. Switching to an opt-in model means that the extra install time is only experienced by users who benefit from Sharp.

### How I propose to do this

I propose a solution that checks for the availability of Sharp at server startup using something like this:

```js
let sharp
 try {
   sharp = require('sharp')
 } catch (e) {
   // Missing-sharp fallback logic
 }
```

We would also do a similar check during SSR or SSG, as part of the default loader provided with the image component. The default loader is what integrates with the image optimizer, so If that loader is called during production, and sharp is not installed, we can log a warning message on the server such as the following:

`WARNING: For production image optimization with Next.js, the optional 'sharp' package is strongly recommended. Run 'yarn add sharp' or install sharp globally, and Next.js will use it automatically for image optimization.`

The Squoosh code will continue to exist in the Next.js codebase, as it is adequate for a development server use case (though even in that context, Sharp will be used if it available).

The Sharp integration code (similar to what was removed in [this PR](https://github.com/vercel/next.js/pull/22253/files?file-filters%5B%5D=.js&file-filters%5B%5D=.json&file-filters%5B%5D=.lock&file-filters%5B%5D=.ts)) will be added back into the codebase, to make use of the Sharp library if it is available.

**Comment by AlexStansfield** [comment]
We experienced a pretty severe degradation in image optimization performance. We're actually looking at 3rd party optimizers to use via loaders instead of the built in optimizer. Would definitely support this option to return to sharp based optimizer instead.

**Comment by Tralgar** [comment]
Hello, I'm experiencing same issue. I've downgraded nextjs but a solution to this huge problem would be nice, your proposal seems to be a good idea ! 
Maybe they are already working on it and try fix this an other way. But for my app it's really a pain, images are so importants... 
Do you think we could fix it alone with sharp and not wait for a nextjs solution ?

**Comment by doptster** [comment]
Hi, we planning to use Next Image in production using Next JS 11, but the first time loading seems to be too slow (at least a few seconds), even using the examples in localhost, it is the same as well. Hopefully we have a resolution for this.

![image](https://user-images.githubusercontent.com/2612333/126729862-cf48a351-b41a-40f6-8656-9cfd0119cf32.png)
![image](https://user-images.githubusercontent.com/2612333/126731323-0d66c621-e5e0-4955-af7b-08af7177892a.png)


Reference: [Image component](https://github.com/vercel/next.js/tree/canary/examples/image-component)

**Comment by karlhorky** [comment]
For those following along, @atcastle has opened a PR that was merged at https://github.com/vercel/next.js/pull/27346 🎉🎉   (already released in `11.0.2-canary.20`)

As for non-canary versions, may be out in `11.0.2`! 🚀

**Reply by styfle** [reply]
You can try it today on the canary channel, `yarn add next@canary sharp`

**Reply by karlhorky** [reply]
@styfle @atcastle any team plans to add `sharp` back as a non-optional dependency? Right now, it makes starting a serious Next.js project 2 commands rather than 1 (`yarn add sharp` in addition to `yarn create next-app ...`)

**Comment by qnxdev** [comment]
The problem is not many people actually know that this is included in 11

**Reply by styfle** [reply]
You should see a warning when running image optimization with `next start`
